src/main/java/org/opensearch/knn/index/query/KNNWeight.java

```diff
@@ -121,54 +124,35 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
         // We should give this condition a deeper look that where it should be placed. For now I feel this is a good
         // place,
         if (filterWeight != null && cardinality == 0) {
-            return Collections.emptyMap();
+            return EMPTY_DOC_IDS_TO_SCORE_MAP;
         }
-
         /*
-         * The idea for this optimization is to get K results, we need to atleast look at K vectors in the HNSW graph
+         * The idea for this optimization is to get K results, we need to at least look at K vectors in the HNSW graph
          * . Hence, if filtered results are less than K and filter query is present we should shift to exact search.
          * This improves the recall.
          */
-        Map<Integer, Float> docIdsToScoreMap;
-        final ExactSearcher.ExactSearcherContext exactSearcherContext = ExactSearcher.ExactSearcherContext.builder()
-            .k(k)
-            .isParentHits(true)
-            .matchedDocs(filterBitSet)
-            // setting to true, so that if quantization details are present we want to do search on the quantized
-            // vectors as this flow is used in first pass of search.
-            .useQuantizedVectorsForSearch(true)
-            .knnQuery(knnQuery)
-            .build();
-        if (filterWeight != null && canDoExactSearch(cardinality)) {
-            docIdsToScoreMap = exactSearch(context, exactSearcherContext);
-        } else {
-            docIdsToScoreMap = doANNSearch(context, filterBitSet, cardinality, k);
-            if (docIdsToScoreMap == null) {
-                return Collections.emptyMap();
-            }
-            if (canDoExactSearchAfterANNSearch(cardinality, docIdsToScoreMap.size())) {
-                log.debug(
-                    "Doing ExactSearch after doing ANNSearch as the number of documents returned are less than "
-                        + "K, even when we have more than K filtered Ids. K: {}, ANNResults: {}, filteredIdCount: {}",
-                    k,
-                    docIdsToScoreMap.size(),
-                    cardinality
-                );
-                docIdsToScoreMap = exactSearch(context, exactSearcherContext);
-            }
+        if (isFilteredExactSearchPreferred(cardinality)) {
+            return doExactSearch(context, acceptedDocs, k);
         }
-        if (docIdsToScoreMap.isEmpty()) {
-            return Collections.emptyMap();
+        Map<Integer, Float> docIdsToScoreMap = doANNSearch(context, acceptedDocs, cardinality, k);
+        if (canDoExactSearch(context, cardinality, docIdsToScoreMap.size())) {
+            // if filter is provided uses filtered docs else use get all docs
+            final BitSet docs = filterWeight != null ? acceptedDocs : getAllDocsBitSet(context);
+            return doExactSearch(context, docs, k);
         }
         return docIdsToScoreMap;
     }
 
+    static {
+        EMPTY_DOC_IDS_TO_SCORE_MAP = Collections.emptyMap();
+    }
```

Comment:
may be move this on top of the class. and why do we even need this? what was the problem with Collections.emptyMap()?

Reply:
Moved to the top. To avoid repeating Collections.emptyMap() in every return statement

Reply:
why have the block and not initialize in declaration?

Reply:
also, are we using this at all?

Reply:
Removed